### PR TITLE
#46 - Added FileStorage test to check that overwrites are supported

### PR DESCRIPTION
**Pull request by olegmoz**

### CODE REVIEW

**Comment by paulodamaso** [comment]
@g4s8 I'm in need of new reviews, please assign me

**Comment by g4s8** [comment]
@0crat assign @paulodamaso

**Comment by 0crat** [comment]
This pull request [#48](https://github.com/artipie/asto/issues/48) is assigned to @iakunin[/z](https://www.0crat.com/u/iakunin), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/51cc44c7-7e65-4faa-a264-e65d36d7a0e3); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/test/java/com/artipie/asto/FileStorageTest.java [hunk 1]
```diff
@@ -93,6 +94,20 @@ public void savesAndLoads() throws Exception {
         );
     }
 
+    @Test
+    public void saveOverwrites() throws Exception {
+        final byte[] original = "1".getBytes();
+        final byte[] updated = "2".getBytes();
+        Assume.assumeThat(updated, Matchers.not(Matchers.equalTo(original)));
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/FileStorageTest.java]: @olegmoz Do not use static matchers. Anyways, this assumption will never be false, so why have it?

File: src/test/java/com/artipie/asto/FileStorageTest.java [hunk 2]
```diff
@@ -93,6 +94,20 @@ public void savesAndLoads() throws Exception {
...
+        final BlockingStorage storage = new BlockingStorage(
+            new FileStorage(this.folder.newFolder().toPath())
+        );
+        final Key key = new Key.From("foo");
+        storage.save(key, original);
+        storage.save(key, updated);
+        MatcherAssert.assertThat(storage.value(key), Matchers.equalTo(updated));
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/FileStorageTest.java]: @olegmoz Do not use static matchers, and put a description to the failure message.

File: src/test/java/com/artipie/asto/FileStorageTest.java [hunk 3]
```diff
@@ -93,6 +94,20 @@ public void savesAndLoads() throws Exception {
         );
     }
 
+    @Test
+    public void saveOverwrites() throws Exception {
+        final byte[] original = "1".getBytes();
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/FileStorageTest.java]: @olegmoz Lets use a constant for this values

File: src/test/java/com/artipie/asto/FileStorageTest.java [hunk 4]
```diff
@@ -93,6 +94,20 @@ public void savesAndLoads() throws Exception {
...
+        final byte[] updated = "2".getBytes();
+        Assume.assumeThat(updated, Matchers.not(Matchers.equalTo(original)));
+        final BlockingStorage storage = new BlockingStorage(
+            new FileStorage(this.folder.newFolder().toPath())
```
Comment (paulodamaso) [inline src/test/java/com/artipie/asto/FileStorageTest.java]: @olegmoz I don't think that it is a good idea to share this between all tests

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by 0crat** [comment]
> @0crat assign @paulodamaso  ([here](https://github.com/artipie/asto/issues/48#issuecomment-584700804))

@g4s8 This pull request [#48](https://github.com/artipie/asto/issues/48) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/eb1cfa47-ab77-4937-8f57-9785796f3865); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
@0crat[/z](https://www.0crat.com/u/0crat) I see that you unassigned this issue; the order is still assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) though; to cancel the order use `refuse`, as in [§6](http://www.zerocracy.com/policy.html#6)

File: src/test/java/com/artipie/asto/FileStorageTest.java (hunk 1 above)
Comment (olegmoz) [reply src/test/java/com/artipie/asto/FileStorageTest.java]: @paulodamaso okey, removed it

File: src/test/java/com/artipie/asto/FileStorageTest.java (hunk 2 above)
Comment (olegmoz) [reply src/test/java/com/artipie/asto/FileStorageTest.java]: @paulodamaso fixed

File: src/test/java/com/artipie/asto/FileStorageTest.java (hunk 3 above)
Comment (olegmoz) [reply src/test/java/com/artipie/asto/FileStorageTest.java]: @paulodamaso there is no reason to bring this value outside of test, it is not going to be used anywhere else

File: src/test/java/com/artipie/asto/FileStorageTest.java (hunk 4 above)
Comment (olegmoz) [reply src/test/java/com/artipie/asto/FileStorageTest.java]: @paulodamaso new folder is created for each new test, so the folder is not being reused

**Comment by olegmoz** [comment]
@paulodamaso made changes by your comments, please review

File: src/test/java/com/artipie/asto/FileStorageTest.java (hunk 1 above)
Comment (paulodamaso) [reply src/test/java/com/artipie/asto/FileStorageTest.java]: @olegmoz Thanks!

File: src/test/java/com/artipie/asto/FileStorageTest.java (hunk 4 above)
Comment (paulodamaso) [reply src/test/java/com/artipie/asto/FileStorageTest.java]: @olegmoz Makes sense.

File: src/test/java/com/artipie/asto/FileStorageTest.java (hunk 3 above)
Comment (paulodamaso) [reply src/test/java/com/artipie/asto/FileStorageTest.java]: @olegmoz Yes, you are right

**Comment by olegmoz** [comment]
@yegor256 please review

**Comment by yegor256** [comment]
@olegmoz there are merge conflicts, please fix them

**Comment by olegmoz** [comment]
@paulodamaso please re-review, merging with master has dismissed your review

**Comment by olegmoz** [comment]
@paulodamaso @g4s8 tests failing are due to bug in RxFile (see #50 ) and is not related to this PR, please merge

**Comment by g4s8** [comment]
@olegmoz we can't break the build in master branch. Let's wait #50 fix to be merged, and then we can merge this PR

**Comment by 0crat** [comment]
@paulodamaso[/z](https://www.0crat.com/u/paulodamaso) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by olegmoz** [comment]
@paulodamaso tests fixed now, please review

**Comment by paulodamaso** [comment]
@g4s8 Merge please

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Pull request merged**